### ISSUE TITLE
Maximum call stack size exceeded - cloning AST with ref cycles

### ISSUE THREAD

**Issue by fatfisz** [rollup/rollup-watch#21]
The error occurs after referencing an unexisting file, then fixing the name so an existing file is referenced.

I've managed to come up with a relatively small example, unfold below.

<details>
Entry file:

```
import React from 'react';

import a from './a.js';

export default <div prop={a} />;
```

File "a.js":

```
export default 42;
```

Rollup config:

```
import babel from 'rollup-plugin-babel';
import commonjs from 'rollup-plugin-commonjs';
import nodeResolve from 'rollup-plugin-node-resolve';


export default {
  entry: 'entry.js',
  dest: 'bundle.js',
  format: 'cjs',
  plugins: [
    nodeResolve(),
    commonjs({
      include: 'node_modules/**',
    }),
    babel({
      exclude: 'node_modules/**',
    }),
  ],
};
```

The output from rollup-watch is:

```
> rollup -c -w

checking rollup-watch version...
bundling...
bundled in 843ms. Watching for changes...

// Here I change the name to something invalid

bundling...
Could not resolve './a-.js' from /some_path/entry.js
Error: Could not resolve './a-.js' from /some_path/entry.js
    at Error (native)
    at /some_path/node_modules/rollup-plugin-node-resolve/dist/rollup-plugin-node-resolve.cjs.js:78:21
    at /some_path/node_modules/resolve/lib/async.js:55:18
    at load (/some_path/node_modules/resolve/lib/async.js:69:43)
    at onex (/some_path/node_modules/resolve/lib/async.js:92:31)
    at /some_path/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:123:15)
Type rollup --help for help, or visit https://github.com/rollup/rollup/wiki

// Here I chang the name back

bundling...
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:24:26)
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:36:18)
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:36:18)
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:36:18)
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:36:18)
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:36:18)
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:36:18)
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:36:18)
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:36:18)
    at deepClone (/some_path/node_modules/rollup/src/utils/object.js:36:18)
Type rollup --help for help, or visit https://github.com/rollup/rollup/wiki
```

</details>

I've started digging around the code a bit, but I don't have much time rn. I've only found out that after the name is restored, [`deepClone`](https://github.com/rollup/rollup/blob/10d2ad36fb308ffdf554b0c41dd27a2f2ed04a69/src/Module.js#L42) receives an AST that has a ref cycle on an inport declaration. A quick check shows that removing `ast ||` from the line above fixes the problem.

Maybe the AST that is cached is modified somewhere later and when it is cloned again the error occurs?

**Comment by fatfisz** [comment]
Btw. no matter what I do after the first "Maximum call stack size exceeded", the same error shows after each change and I have to restart.

**Comment by dev-nima** [comment]
i have the same issue when i use version 2.5.0

**Comment by pascalwhoop** [comment]
same here, just found this because of the Ionic RC0

```
[19:49:26]  RangeError: Maximum call stack size exceeded
    at isArray (native)
    at deepClone (/Users/me/Demployeruments/Code/employer/project/frontend_tablet/node_modules/rollup/dist/rollup.js:157:7)
    at deepClone (/Users/me/Demployeruments/Code/employer/project/frontend_tablet/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/Users/me/Demployeruments/Code/employer/project/frontend_tablet/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/Users/me/Demployeruments/Code/employer/project/frontend_tablet/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/Users/me/Demployeruments/Code/employer/project/frontend_tablet/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/Users/me/Demployeruments/Code/employer/project/frontend_tablet/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/Users/me/Demployeruments/Code/employer/project/frontend_tablet/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/Users/me/Demployeruments/Code/employer/project/frontend_tablet/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/Users/me/Demployeruments/Code/employer/project/frontend_tablet/node_modules/rollup/dist/rollup.js:165:18)
```

**Comment by sunjay** [comment]
Same issue using version 2.5.0

As described, `deepClone` is called a bunch of times and then rollup dies. I have to restart the watch process whenever it happens.

**Comment by jonataswalker** [comment]
> Same issue using version 2.5.0

**Comment by Rich-Harris** [comment]
The bug was in Rollup itself – fixed (finally!) in 0.39.2
